src/server/worker/worker.go

```diff
@@ -1,29 +1,198 @@
-	// line's inputs, or it's filtered
-	matchesData := true
-dataFilters:
-	for _, dataFilter := range filter {
-		for _, datum := range data {
-			if dataFilter == datum.Path ||
-				dataFilter == base64.StdEncoding.EncodeToString(datum.Hash) ||
-				dataFilter == hex.EncodeToString(datum.Hash) {
-				continue dataFilters // Found, move to next filter
+const (
+	masterLockPath = "_master_worker_lock"
+)
+
+// The Worker object represents
+type Worker struct {
+	APIServer *server.APIServer // Provides rpcs for other nodes in the cluster
+	driver    driver.Driver     // Provides common functions used by worker code
+	namespace string            // The namespace in which pachyderm is deployed - TODO: what is this for?
+	status    *transform.Status // An interface for inspecting and canceling the actively running task
+}
+
+// NewWorker constructs a Worker object that provides all worker functionality:
+//  1. a master goroutine that attempts to obtain the master lock for the pipeline workers and direct jobs
+//  2. a worker goroutine that gets tasks from the master and processes them
+//  3. an api server that serves requests for status or cross-worker communication
+//  4. a driver that provides common functionality between the above components
+func NewWorker(
+	pachClient *client.APIClient,
+	etcdClient *etcd.Client,
+	etcdPrefix string,
+	pipelineInfo *pps.PipelineInfo,
+	workerName string,
+	namespace string,
+	hashtreePath string,
+) (*Worker, error) {
+	stats.InitPrometheus()
+
+	hasDocker := true
+	if _, err := os.Stat("/var/run/docker.sock"); err != nil {
+		hasDocker = false
+	}
+
+	driver, err := driver.NewDriver(
+		pipelineInfo,
+		pachClient,
+		etcdClient,
+		etcdPrefix,
+		filepath.Join(hashtreePath, uuid.NewWithoutDashes()),
+		client.PPSInputPrefix,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if pipelineInfo.Transform.Image != "" && hasDocker {
+		docker, err := docker.NewClientFromEnv()
+		if err != nil {
+			return nil, err
+		}
+		image, err := docker.InspectImage(pipelineInfo.Transform.Image)
+		if err != nil {
+			return nil, fmt.Errorf("error inspecting image %s: %+v", pipelineInfo.Transform.Image, err)
+		}
+		if pipelineInfo.Transform.User == "" {
+			pipelineInfo.Transform.User = image.Config.User
+		}
+		if pipelineInfo.Transform.WorkingDir == "" {
+			pipelineInfo.Transform.WorkingDir = image.Config.WorkingDir
+		}
+		if pipelineInfo.Transform.Cmd == nil {
+			if len(image.Config.Entrypoint) == 0 {
+				ppsutil.FailPipeline(pachClient.Ctx(), etcdClient, driver.Pipelines(),
+					pipelineInfo.Pipeline.Name,
+					"nothing to run: no transform.cmd and no entrypoint")
 			}
+			pipelineInfo.Transform.Cmd = image.Config.Entrypoint
+		}
+	}
+
+	worker := &Worker{
+		driver:    driver,
+		namespace: namespace,
+		status:    &transform.Status{},
+	}
+
+	worker.APIServer = server.NewAPIServer(driver, worker.status, workerName)
+
+	go worker.master(etcdClient, etcdPrefix)
+	go worker.worker()
```

Comment:
The entry points for the master and worker goroutines are here (you can consider this the new `worker/api_server.go` and `worker/master.go`), but most of the underlying functionality is outsourced to smaller-scoped packages.